panic: not a collection type (regression between 13.1 and 13.0)

# Terraform Version
```
Terraform v0.13.1
+ provider registry.terraform.io/hashicorp/azurerm v2.24.0
```

### Terraform Configuration Files
```terraform
terraform {
  backend "azurerm" {}
}
provider "azurerm" {
  version = "~>2"
  features {}
  subscription_id = var.subscription_id
}
data "azurerm_client_config" "current" {}
data "terraform_remote_state" "geography" {
  for_each = var.geos
  backend = "azurerm"
  config = {
    resource_group_name = var.backend_state.resource_group_name
    storage_account_name = var.backend_state.account_name
    container_name = var.backend_state.container_name
    key = "infra-setup-${each.value}.tfstate"
  }
}
data "terraform_remote_state" "locust" {
  count = var.locust_supported ? 1 : 0
  backend = "azurerm"
  config = {
    resource_group_name = var.backend_state.resource_group_name
    storage_account_name = var.backend_state.account_name
    container_name = var.backend_state.container_name
    key = "infra-locust.tfstate"
  }
}
locals {
  app_gateway_object_ids = flatten(values({
    for item in flatten([
      for geo in var.geos : [
        for object_id in values(data.terraform_remote_state.geography[geo].outputs.app_gateway_msi_objectids) : {
          key = "${geo}_${object_id}"
          policy = {
            object_id = object_id
            key_permissions = [ ]
            secret_permissions = [ "get" ]
            certificate_permissions = [ ]
          }
        }
      ]
    ]) : item.key => item.policy
  }))
  locust_app_gateway_policies = flatten(values({
    for item in flatten([
      for num in range(var.locust_supported ? 1 : 0) :[
        for object_id in values(data.terraform_remote_state.locust[num].outputs.app_gateway_msi_objectids) : {
          key = "${num}_${object_id}"
          policy = {
            object_id = object_id
            key_permissions = [ ]
            secret_permissions = [ "get" ]
            certificate_permissions = [ ]
          }
        }
      ]
    ]) : item.key => item.policy
  }))
}
resource "azurerm_resource_group" "rg" {
  name = var.resource_group_name
  location = "westus2"
}
resource "azurerm_user_assigned_identity" "msi" {
  resource_group_name = azurerm_resource_group.rg.name
  location = azurerm_resource_group.rg.location
  name = var.msi_name
}
resource "azurerm_role_assignment" "managed_identity_operator" {
  scope = azurerm_user_assigned_identity.msi.id
  role_definition_name = "Managed Identity Operator"
  principal_id = var.cluster_service_principal_object_id
}
module "keyvault" {
  source = "../modules/keyvault"
  resource_group_name = azurerm_resource_group.rg.name
  account_name = var.key_vault_name
  location = azurerm_resource_group.rg.location
  tenant_id = data.azurerm_client_config.current.tenant_id
  enabled_for_deployment = true
  enabled_for_template_deployment = true
  policies = concat(
    var.policies,
    [
      {
        object_id = azurerm_user_assigned_identity.msi.principal_id
        key_permissions = [ ]
        secret_permissions = [ "get", "list" ]
        certificate_permissions = [ ]
      }
    ],
    flatten(local.app_gateway_object_ids),
    flatten(local.locust_app_gateway_policies))
}
output "key_vault_id" {
  value = module.keyvault.id
}
output "key_vault_uri" {
  value = module.keyvault.vault_uri
}
output "service_msi" {
  value = {
    resource_id = azurerm_user_assigned_identity.msi.id
    client_id = azurerm_user_assigned_identity.msi.client_id
    msi_name = var.msi_name
  }
}
```

### Debug Output
```
2020/08/26 14:25:52 [TRACE] Executing graph transform *terraform.OrphanResourceInstanceCountTransformer
2020/08/26 14:25:52 [TRACE] eval: *terraform.EvalSequence
2020/08/26 14:25:52 [TRACE] eval: *terraform.EvalGetProvider
2020/08/26 14:25:52 [TRACE] Completed graph transform *terraform.OrphanResourceInstanceCountTransformer (no changes)
2020/08/26 14:25:52 [TRACE] Executing graph transform *terraform.AttachStateTransformer
2020/08/26 14:25:52 [TRACE] Completed graph transform *terraform.AttachStateTransformer (no changes)
2020/08/26 14:25:52 [TRACE] Executing graph transform *terraform.TargetsTransformer
2020/08/26 14:25:52 [TRACE] Completed graph transform *terraform.TargetsTransformer (no changes)
2020/08/26 14:25:52 [TRACE] Executing graph transform *terraform.ReferenceTransformer
2020/08/26 14:25:52 [TRACE] eval: *terraform.EvalReadState
2020/08/26 14:25:52 [TRACE] EvalReadState: reading state for azurerm_resource_group.rg
2020/08/26 14:25:52 [DEBUG] ReferenceTransformer: "data.azurerm_client_config.current" references: []
2020/08/26 14:25:52 [TRACE] Completed graph transform *terraform.ReferenceTransformer (no changes)
2020/08/26 14:25:52 [TRACE] Executing graph transform *terraform.RootTransformer
2020/08/26 14:25:52 [TRACE] Completed graph transform *terraform.RootTransformer (no changes)
2020/08/26 14:25:52 [TRACE] vertex "data.azurerm_client_config.current": entering dynamic subgraph
2020/08/26 14:25:52 [TRACE] dag/walk: visiting "data.azurerm_client_config.current"
2020/08/26 14:25:52 [TRACE] vertex "data.azurerm_client_config.current": starting visit (*terraform.NodePlannableResourceInstance)
2020/08/26 14:25:52 [TRACE] vertex "data.azurerm_client_config.current": evaluating
2020/08/26 14:25:52 [TRACE] [walkPlan] Entering eval tree: data.azurerm_client_config.current
2020/08/26 14:25:52 [TRACE] eval: *terraform.EvalSequence
2020/08/26 14:25:52 [TRACE] eval: *terraform.EvalGetProvider
2020/08/26 14:25:52 [TRACE] UpgradeResourceState: schema version of azurerm_resource_group.rg is still 0; calling provider "azurerm" for any other minor fixups
2020/08/26 14:25:52 [TRACE] eval: *terraform.EvalReadState
2020/08/26 14:25:52 [TRACE] GRPCProvider: UpgradeResourceState
2020/08/26 14:25:52 [TRACE] EvalReadState: reading state for data.azurerm_client_config.current
2020/08/26 14:25:52 [TRACE] eval: *terraform.EvalValidateSelfRef
2020/08/26 14:25:52 [TRACE] eval: *terraform.evalReadDataPlan
2020/08/26 14:25:52 [TRACE] vertex "data.azurerm_client_config.current": visit complete
panic: not a collection type
...
```

### Crash Output
```
2020/08/26 14:25:52 [TRACE] vertex "data.azurerm_client_config.current": visit complete
panic: not a collection type

goroutine 459 [running]:
github.com/zclconf/go-cty/cty.Type.ElementType(0x2d39220, 0xc000ff80d0, 0xc0005f59e8, 0x8)
	/home/circleci/project/project/vendor/github.com/zclconf/go-cty/cty/collection.go:25 +0xc1
github.com/hashicorp/terraform/terraform.createEmptyBlocks(0xc000f3a8a0, 0x2d39220, 0xc000ff80f0, 0x2355700, 0xc001508ed0, 0x2d39220, 0xc000ff81f0, 0x2355700, 0xc00152cd20)
	/home/circleci/project/project/terraform/eval_read_data_plan.go:242 +0x209
github.com/hashicorp/terraform/terraform.dataObjectHasChanges(0xc000f3a8a0, 0x2d39220, 0xc000ff80f0, 0x2355700, 0xc001508ed0, 0x2d39220, 0xc000ff81f0, 0x2355700, 0xc00152cd20, 0x0, ...)
	/home/circleci/project/project/terraform/eval_read_data_plan.go:211 +0x1a5
github.com/hashicorp/terraform/terraform.(*evalReadDataPlan).Eval(0xc001516000, 0x2d7b6e0, 0xc0012121a0, 0x1, 0x1, 0x0, 0x0)
	/home/circleci/project/project/terraform/eval_read_data_plan.go:107 +0x123f
github.com/hashicorp/terraform/terraform.EvalRaw(0x2cd1460, 0xc001516000, 0x2d7b6e0, 0xc0012121a0, 0x0, 0x0, 0x0, 0x0)
	/home/circleci/project/project/terraform/eval.go:49 +0xce
github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc000692580, 0x2d7b6e0, 0xc0012121a0, 0x1, 0x1, 0x6, 0xc000765550)
	/home/circleci/project/project/terraform/eval_sequence.go:20 +0x104
github.com/hashicorp/terraform/terraform.EvalRaw(0x2cd0ec0, 0xc000692580, 0x2d7b6e0, 0xc0012121a0, 0x230aac0, 0x40098a2, 0x226ae60, 0xc0012f2730)
	/home/circleci/project/project/terraform/eval.go:49 +0xce
github.com/hashicorp/terraform/terraform.Eval(0x2cd0ec0, 0xc000692580, 0x2d7b6e0, 0xc0012121a0, 0xc000692580, 0x2cd0ec0, 0xc000692580, 0x0)
	/home/circleci/project/project/terraform/eval.go:35 +0x54
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x2699c60, 0xc0012f2250, 0x0, 0x0, 0x0)
	/home/circleci/project/project/terraform/graph.go:73 +0xc7e
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc000765440, 0x2699c60, 0xc0012f2250, 0xc0014b2280)
	/home/circleci/project/project/dag/walk.go:387 +0x35e
created by github.com/hashicorp/terraform/dag.(*Walker).Update
	/home/circleci/project/project/dag/walk.go:309 +0x11de
```

### Expected Behavior
Running the following command should succeed.

### Actual Behavior
Crashes on 13.1, but doesn't on 13.0

### Steps to Reproduce
1. `az login`
2. `az account set -s <id>`
3. `terraform init -backend-config=storage_account_name=x -backend-config=container_name=state -backend-config=key=y-bootstrap.tfstate -backend-config=resource_group_name=z-state -backend-config=arm_subscription_id=subid`
4. `terraform plan --var-file ./dev.tfvars -out ./dev.tfplan`

### References
* May be related to #25773

Get this error with Azure AD provider when using 13.1
`data.azuread_users.users configuration is complete, so reading from provider
2020/08/27 11:42:53 [TRACE] GRPCProvider: ReadDataSource
panic: not a collection type`

--- a/terraform/eval_read_data_plan.go
+++ b/terraform/eval_read_data_plan.go
@@ -239,8 +239,6 @@ func createEmptyBlocks(schema *configschema.Block, val cty.Value) cty.Value {
 			continue
 		}
 
-		ety := block.Type().ElementType()
-
 		// helper to build the recursive block values
 		nextBlocks := func() []cty.Value {
 			// this is only called once we know this is a non-null List or Set
@@ -259,6 +257,7 @@ func createEmptyBlocks(schema *configschema.Block, val cty.Value) cty.Value {
 		// We are only concerned with block types that can come from the legacy
 		// sdk, which means TypeList or TypeSet.
 		case configschema.NestingList:
+			ety := block.Type().ElementType()
 			switch {
 			case block.IsNull():
 				objMap[name] = cty.ListValEmpty(ety)
@@ -269,6 +268,7 @@ func createEmptyBlocks(schema *configschema.Block, val cty.Value) cty.Value {
 			}
 
 		case configschema.NestingSet:
+			ety := block.Type().ElementType()
 			switch {
 			case block.IsNull():
 				objMap[name] = cty.SetValEmpty(ety)

--- a/terraform/eval_read_data_plan_test.go
+++ b/terraform/eval_read_data_plan_test.go
@@ -86,6 +86,22 @@ func TestReadDataCreateEmptyBlocks(t *testing.T) {
 		},
 	}
 
+	singleSchema := &configschema.Block{
+		BlockTypes: map[string]*configschema.NestedBlock{
+			"single": {
+				Nesting: configschema.NestingSingle,
+				Block: configschema.Block{
+					Attributes: map[string]*configschema.Attribute{
+						"attr": {
+							Type:     cty.String,
+							Optional: true,
+						},
+					},
+				},
+			},
+		},
+	}
+
 	for _, tc := range []struct {
 		name   string
 		schema *configschema.Block
@@ -328,6 +344,20 @@ func TestReadDataCreateEmptyBlocks(t *testing.T) {
 						),
 					}),
 				}),
+			}),
+		},
+		{
+			"single-block-null",
+			singleSchema,
+			cty.ObjectVal(map[string]cty.Value{
+				"single": cty.NullVal(cty.Object(map[string]cty.Type{
+					"attr": cty.String,
+				})),
+			}),
+			cty.ObjectVal(map[string]cty.Value{
+				"single": cty.NullVal(cty.Object(map[string]cty.Type{
+					"attr": cty.String,
+				})),
 			}),
 		},
 	} {

